components/webln.js

```diff
@@ -0,0 +1,106 @@
+import { createContext, createRef, useContext, useEffect, useImperativeHandle, useState } from 'react'
+
+const WebLNContext = createContext({})
+export const WebLNContextRef = createRef()
+
+const fetchWebLNProvider = async () => {
+  // sync provider from local storage
+  // TODO check if there is a better, more secure option
+  const configJson = window.localStorage.getItem('bc:config')
+  if (configJson) {
+    const WebLNProviders = await import('@getalby/bitcoin-connect').then((mod) => mod.WebLNProviders)
+    const config = JSON.parse(configJson)
+    if (config.connectorType === 'lnbits') return new WebLNProviders.LnbitsWebLNProvider(config.lnbitsInstanceUrl, config.lnbitsAdminKey)
+    // TODO implement other connector types
+  }
+}
+
+export function WebLNProvider ({ children }) {
+  const [provider, setProvider] = useState(null)
+  const [info, setInfo] = useState(null)
+  const [balance, setBalance] = useState(0)
+
+  const initProvider = async (provider) => {
+    const WebLNProviders = await import('@getalby/bitcoin-connect').then((mod) => mod.WebLNProviders)
+    if (provider instanceof WebLNProviders.LnbitsWebLNProvider) {
+      const connector = 'lnbits'
+      // TODO
+      //   we're reaching into private parts here.
+      //   is there a better way to show to user which connection was used?
+      const { hostname: connection } = new URL(provider._instanceUrl)
+      setInfo(o => ({ ...o, connector, connection }))
+    }
+    setProvider(provider)
+    const info = await provider.getInfo()
+    setInfo(o => ({ ...o, ...info }))
+    const { balance } = await provider.getBalance()
+    setBalance(balance)
+  }
+
+  const clearProvider = () => {
+    setProvider(null)
+    setInfo(null)
+  }
+
+  useEffect(() => {
+    async function effect () {
+      const provider = await fetchWebLNProvider()
+      if (!provider) return
+      initProvider(provider)
+    }
+    effect().catch(console.error)
+  }, [])
+
+  // event listeners
+  useEffect(() => {
+    let onConnectedUnsub
+    let onDisconnectedUnsub
+    async function effect () {
+      const onConnected = await import('@getalby/bitcoin-connect-react').then((mod) => mod.onConnected)
+      onConnectedUnsub = onConnected(async (provider) => {
+        // TODO
+        //   why is this not fired on page load when there is a provider available in local storage?
+        //   why do i have to do this myself?
+        // console.log('webln:: connected', provider)
+        // TODO do we need this?
+        // window.webln = provider
+        initProvider(provider)
+      })
+      const onDisconnected = await import('@getalby/bitcoin-connect-react').then((mod) => mod.onDisconnected)
+      onDisconnectedUnsub = onDisconnected(() => {
+        // console.log('webln:: disconnected')
+        clearProvider()
+      })
+    }
+    effect().catch(console.error)
+    return () => {
+      onConnectedUnsub?.()
+      onDisconnectedUnsub?.()
+    }
+  }, [])
+
+  // poll balance
```

Comment:
maybe rather than polling you could just get the balance again after a payment is made? (this won't account for external payments, but I don't really think it is an issue)

Reply:
> maybe rather than polling you could just get the balance again after a payment is made?

Ideally, I don't need to poll at all. Is there a way that the connection can tell me when the balance was updated?

> (this won't account for external payments, but I don't really think it is an issue)

Well, I think it would be an issue :smile: We want to show the latest balance on SN, not something that might be stale. We will also probably need to know the _current_ balance for other reasons like [paying from both wallets at the same time.](https://github.com/stackernews/stacker.news/pull/715#discussion_r1437531356)

Reply:
Ignore this, I think we don't need `get_balance` anymore. I've changed my mind after reading the discussion in https://github.com/MutinyWallet/mutiny-node/issues/895. See [this comment](https://github.com/stackernews/stacker.news/pull/715#discussion_r1437934087) from me.